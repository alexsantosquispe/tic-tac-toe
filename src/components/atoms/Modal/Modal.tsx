import { AnimatePresence, motion } from 'framer-motion';

import type { ReactNode } from 'react';
import { useTranslation } from 'react-i18next';
import usePressKey from '../../../hooks/usePressKey';
import { XIcon } from '../../../icons/XIcon';

interface ModalProps {
  isOpen: boolean;
  title: string;
  children: ReactNode;
  onClose: () => void;
}

export const Modal = ({ isOpen, title, children, onClose }: ModalProps) => {
  const { t } = useTranslation();
  usePressKey('Escape', onClose, !isOpen);

  return (
    <AnimatePresence>
      {isOpen && (
        <motion.section
          data-testid="modal"
          key="backdrop"
          initial={{ opacity: 0 }}
          animate={{ opacity: 1 }}
          exit={{ opacity: 0 }}
          transition={{ duration: 0.1 }}
          className="fixed z-50 flex h-full w-full items-center justify-center bg-black/50 p-8 backdrop-blur-xs dark:bg-black/20"
        >
          <motion.div
            key="modal"
            initial={{ opacity: 0, scale: 0.9, y: 20 }}
            animate={{ opacity: 1, scale: 1, y: 0 }}
            exit={{ opacity: 0, scale: 0.9, y: 20 }}
            transition={{ duration: 0.1, ease: 'easeOut' }}
            className="flex w-full flex-col rounded-2xl bg-white md:w-80 dark:border dark:border-white/20 dark:bg-neutral-950"
          >
            <button
              onClick={onClose}
              aria-label={t('closeModalButtonAriaLabel')}
              className="mt-5 mr-5 self-end rounded-xl text-neutral-500 hover:cursor-pointer hover:text-black dark:text-neutral-600 dark:hover:text-white"
            >
              <XIcon />
            </button>

            <h2 className="text-center text-xl font-medium md:text-2xl">
              {title}
            </h2>

            <div className="flex w-full">{children}</div>
          </motion.div>
        </motion.section>
      )}
    </AnimatePresence>
  );
};
