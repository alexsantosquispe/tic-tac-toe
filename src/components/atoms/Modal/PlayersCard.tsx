import type { ReactNode } from 'react';
import { BotIcon } from '../../../icons/BotIcon';

interface PlayersCardProps {
  isSinglePlayer?: boolean;
  playerIcon: ReactNode;
}

export const PlayersCard = ({
  isSinglePlayer,
  playerIcon
}: PlayersCardProps) => {
  return (
    <div className="flex items-center justify-between gap-8 px-8 py-4">
      {playerIcon}
      <span className="text-xl font-medium">VS</span>
      {isSinglePlayer ? <BotIcon className="size-12" /> : playerIcon}
    </div>
  );
};
