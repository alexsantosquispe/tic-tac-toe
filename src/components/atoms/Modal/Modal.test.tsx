import '@testing-library/jest-dom';

import { fireEvent, render, screen } from '@testing-library/react';

import { I18NWrapper } from '../../../tests/testsUtils';
import { Modal } from './Modal';

describe('Modal', () => {
  const onCloseMock = jest.fn();
  const props = {
    isOpen: true,
    title: 'Modal Title',
    children: <div className="flex p-8">Modal Content</div>,
    onClose: onCloseMock
  };

  describe('styles', () => {
    it('should render the component correctly', () => {
      const component = render(
        <I18NWrapper>
          <Modal {...props} />
        </I18NWrapper>
      );

      expect(component).toMatchSnapshot();

      expect(screen.getByText(props.title)).toBeInTheDocument();

      expect(screen.getByText('Modal Content')).toBeInTheDocument();
    });
  });

  describe('behavior', () => {
    beforeEach(() => {
      render(
        <I18NWrapper>
          <Modal {...props} />
        </I18NWrapper>
      );
    });

    it('should call the onClose function when the close button is clicked', () => {
      const closeButton = screen.getByRole('button');

      expect(closeButton).toBeInTheDocument();

      fireEvent.click(closeButton);

      expect(onCloseMock).toHaveBeenCalledTimes(1);
    });

    it('should call the onClose function when Escape key is pressed', () => {
      fireEvent.keyDown(document, { key: 'Escape' });

      expect(onCloseMock).toHaveBeenCalledTimes(1);
    });
  });
});
