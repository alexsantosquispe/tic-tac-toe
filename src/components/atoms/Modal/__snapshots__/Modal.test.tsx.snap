// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Modal should render the component correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="fixed z-50 flex h-full w-full items-center justify-center bg-black/50 p-8 backdrop-blur-xs dark:bg-black/20"
        style="opacity: 0;"
      >
        <div
          class="flex w-full flex-col rounded-2xl bg-white md:w-80 dark:border dark:border-white/20 dark:bg-neutral-950"
          style="opacity: 0; transform: translateY(20px) scale(0.9);"
        >
          <button
            aria-label="Close modal button"
            class="mt-5 mr-5 self-end rounded-xl text-neutral-500 hover:cursor-pointer hover:text-black dark:text-neutral-600 dark:hover:text-white"
          >
            <svg
              class="size-6"
              data-testid="x-icon"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="3"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M18 6 6 18"
              />
              <path
                d="m6 6 12 12"
              />
            </svg>
          </button>
          <h2
            class="text-center text-xl font-medium md:text-2xl"
          >
            Modal Title
          </h2>
          <div
            class="flex w-full"
          >
            <div
              class="flex p-8"
            >
              Modal Content
            </div>
          </div>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="fixed z-50 flex h-full w-full items-center justify-center bg-black/50 p-8 backdrop-blur-xs dark:bg-black/20"
      style="opacity: 0;"
    >
      <div
        class="flex w-full flex-col rounded-2xl bg-white md:w-80 dark:border dark:border-white/20 dark:bg-neutral-950"
        style="opacity: 0; transform: translateY(20px) scale(0.9);"
      >
        <button
          aria-label="Close modal button"
          class="mt-5 mr-5 self-end rounded-xl text-neutral-500 hover:cursor-pointer hover:text-black dark:text-neutral-600 dark:hover:text-white"
        >
          <svg
            class="size-6"
            data-testid="x-icon"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="3"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18 6 6 18"
            />
            <path
              d="m6 6 12 12"
            />
          </svg>
        </button>
        <h2
          class="text-center text-xl font-medium md:text-2xl"
        >
          Modal Title
        </h2>
        <div
          class="flex w-full"
        >
          <div
            class="flex p-8"
          >
            Modal Content
          </div>
        </div>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
